src/web-server.js

```diff
@@ -118,54 +120,102 @@ app.patch("/import-map.json", (req, res) => {
-      res
+  if (req.body.scopes) {
+    if (getConfig().manifestFormat !== "importmap") {
+      return res
         .status(400)
         .send(
-          `Invalid import map in request body -- module with name '${moduleName}' does not have a string url`
+          `Invalid import map in request body -- scopes are only supported with manifest format "importmap"`
         );
-      return;
+    }
+
+    if (typeof req.body.scopes !== "object") {
+      return res
+        .status(400)
+        .send(`Invalid import map in request body -- scopes is not an object`);
+    }
+
+    for (let scopeName in req.body.scopes) {
+      if (typeof req.body.scopes[scopeName] !== "object") {
+        return res
+          .status(400)
+          .send(
+            `Invalid import map in request body -- scope with name '${scopeName}' is not an object`
+          );
+      }
+
+      if (Object.keys(req.body.scopes[scopeName]).length === 0) {
+        return res
+          .status(400)
+          .send(
+            `Invalid import map in request body -- scope with name '${scopeName}' is an object with no properties`
+          );
+      }
     }
   }
 
-  const importUrls = Object.values(req.body.imports);
+  let validImportUrlPromises = Promise.resolve();
+  if (req.body.imports) {
+    // Confirm the imports are working
+    const importUrls = Object.values(req.body.imports);
+    const unsafeUrls = importUrls.map(checkUrlUnsafe).filter(Boolean);
 
-  const unsafeUrls = importUrls.map(checkUrlUnsafe).filter(Boolean);
+    if (unsafeUrls.length > 0) {
+      return res.status(400).send({
+        error: `The following URLs are not trusted - ${unsafeUrls.join(", ")}`,
+      });
+    }
 
-  if (unsafeUrls.length > 0) {
-    return res.status(400).send({
-      error: `The following URLs are not trusted - ${unsafeUrls.join(", ")}`,
-    });
+    validImportUrlPromises = importUrls.map((url) => verifyValidUrl(req, url));
   }
 
-  const validImportUrlPromises = importUrls.map((url) =>
-    verifyValidUrl(req, url)
-  );
+  let validScopeUrlPromises = Promise.resolve();
+  if (req.body.scopes) {
+    const scopeUrlsToValidate = findUrlsToValidateInScopes(req.body.scopes);
+    const unsafeUrls = scopeUrlsToValidate.map(checkUrlUnsafe).filter(Boolean);
+
+    if (unsafeUrls.length > 0) {
+      return res.status(400).send({
+        error: `The following URLs are not trusted - ${unsafeUrls.join(", ")}`,
+      });
+    }
+
+    if (scopeUrlsToValidate.length > 0) {
+      validScopeUrlPromises = scopeUrlsToValidate.map((url) =>
+        verifyValidUrl(req, url)
+      );
+    }
+  }
 
-  Promise.all(validImportUrlPromises)
+  return Promise.all([validImportUrlPromises, validScopeUrlPromises])
     .then(() => {
-      modify
-        .modifyMultipleServices(env, req.body.imports)
+      Promise.all([
+        modify.modifyMultipleServices(env, req.body.imports),
+        modify.modifyMultipleScopes(env, req.body.scopes),
```

Comment:
There is a problem with this code - it modifies the `imports` and `scopes` as separate "modify locks." This means that the work happens sequentially, and we write the import map to the target destination twice. This has the chance for bugs, since updating either the imports or scopes first without simultaneously updating the other could result in the deployed import map having a bad state. That bad state would usually exist only for less than a second, but even that period of time is problematic for a production environment. Additionally, if the first modification succeeds but the second one fails, the import map would remain in the bad state indefinitely. Updating the import map should be atomic - should all succeed or fail together.

In other words, this code will cause a production environment to have the wrong import map.

To fix, we should modify both the services and the scopes all in the same modifyLock. This code will need to be refactored to do so.